src/EditorFeatures/CSharpTest/AutomaticCompletion/AutomaticLineEnderTests.cs

```diff
@@ -845,12 +845,11 @@ protected override Action CreateNextHandler(TestWorkspace workspace)
             return () => { };
         }
 
-        internal override ICommandHandler<AutomaticLineEnderCommandArgs> CreateCommandHandler(
-            Microsoft.CodeAnalysis.Editor.Host.IWaitIndicator waitIndicator,
+        internal override IChainedCommandHandler<AutomaticLineEnderCommandArgs> CreateCommandHandler(
             ITextUndoHistoryRegistry undoRegistry,
             IEditorOperationsFactoryService editorOperations)
         {
-            return new AutomaticLineEnderCommandHandler(waitIndicator, undoRegistry, editorOperations);
```

Comment:
how does wait handling work in the new system?

Reply:
1. IWaitIndicator/IWaitContext are platformazed and public (need better names though - we end up with IWaitableUIOperationContext kind of name). I'll use old names to make simpler to explain.
2. All command handlers are executed under a shared wait indicator/context set up by the editor and can obtain wait context via CommandExecutionContext argument.
4. With the wait context command handlers get shared cancellation token.
5. Because the wait context is shared between command handlers, I added a concept of "wait context scope". When a command handler needs to affect the shared wait context (allow or disallow cancellation, add its own message or progress) it pushes a new wait context scope and disposes it when it's done. Shared wait context deals with aggregating these scopes into the wait dialog. Cancellability is aggregated in a sticky way though - if one handler pushed a scope to disallow cancellation and didn't just handle the command, it stays disabled even after the handler is done.

Reply:
@CyrusNajmabadi I linked the spec in description and did a bunch of improvements. Please take another look when you have time. Thanks!

Reply:
@CyrusNajmabadi 
Regarding additional 3000 lines: most of them are coming from new localized command handler names in xlf files (25 handlers x 10 lines per new string x number of languages).

Regarding chaining - I'm going to disappoint you but we gave up on  pure non-chained approach and ended up with dual model (simple and if needed - chained). See spec for design consideration.

Reply:
>Regarding chaining - I'm going to disappoint you but we gave up on pure non-chained approach and ended up with dual model (simple and if needed - chained)

That doesn't disappoint me :)  I'm not actually surprised given the complexity of some of our chains.